### ISSUE TITLE
No Julia functions inside GAT signatures

### ISSUE THREAD

**Issue by epatters** [epatters/Catlab.jl#112]
Since I originally thought of signatures like Haskell type classes, arbitrary Julia function definitions are allowed in signatures of GATs, e.g., for defining Unicode syntax. I would like to ban this and draw a clear distinction between the algebraic theories and the Julia code that instantiates them. This would be a first step towards a better internal representation of GATs (cf. #15).

**Comment by jpfairbanks** [comment]
Is this closed by the recent work on the Unicode syntax such as #129?

**Comment by epatters** [comment]
Not yet, although the groundwork has been laid. At this point, I think @mehalter could resolve this issue fairly quickly.

**Comment by epatters** [comment]
Here is some background info and a suggested approach, based on issues that Micah noticed in his PR #130.

The original intent behind allowing function definitions in signatures, besides supporting Unicode operators, was to allow certain term constructors to have default implementations via other terms. For example, in the definition of `CompactClosedCategory`, we currently see this:

```julia
hom(A::Ob, B::Ob) = B ⊗ dual(A)
ev(A::Ob, B::Ob) = id(B) ⊗ (σ(dual(A), A) ⋅ dcounit(A))
curry(A::Ob, B::Ob, f::Hom) = ((id(A) ⊗ (dunit(B) ⋅ σ(dual(B), B))) ⋅ (f ⊗ id(dual(B))))
```

This expresses the fact that the closed monoidal structure of a compact closed category is derived from its dual objects. However, as argued above, this conflates the algebraic theory with its Julia implementation. Thanks to Micah, we can now record axioms, so the function definitions above can be replaced with equations:

```julia
(hom(A,B) == B ⊗ dual(A)) where (A::Ob, B::Ob)
(ev(A,B) == id(B) ⊗ (σ(dual(A), A) ⋅ dcounit(A))) where (A::Ob, B::Ob)
(curry(A, B, f) == ((id(A) ⊗ (dunit(B) ⋅ σ(dual(B), B))) ⋅ (f ⊗ id(dual(B))))) where (A::Ob, B::Ob, C::Ob, f::Hom(A⊗B,C))
```

Another reason that such definitions should be handled by axioms is that sometimes there are sets of terms that can each be defined via the others. For example, in `CartesianCategory`, we have

```julia
pair(f,g) == compose(mcopy(A), otimes(f,g)) where (A::Ob, B::Ob, C::Ob, f::Hom(A,B), g::Hom(A,C))
proj1(A,B) == otimes(id(A), delete(B)) where (A::Ob, B::Ob)
proj2(A,B) == otimes(delete(A), id(B)) where (A::Ob, B::Ob)
```

defining the pairing and projections in terms of copying and deleting, but conversely the copying and deleting can be defined in terms of pairing and projection as

```julia
mcopy(A) == pair(id(A), id(A)) where A::Ob
delete(A) == proj1(id(munit()), A) where A::Ob
delete(A) == proj2(A, id(munit())) where A::Ob # Redundant but true
```

Thus, the question of which terms are taken to be primitive or derived should be left to particular syntax systems.

**Comment by epatters** [comment]
For now, we can translate function definitions in signatures into axioms and, if desired or expected by existing code, add them to the default syntax systems. Eventually, we may want to add a special form of "definitional" axiom, e.g.

```julia
pair(f,g) := compose(mcopy(A), otimes(f,g)) where (A::Ob, B::Ob, C::Ob, f::Hom(A,B), g::Hom(A,C))
```

which, in addition to establishing the usual equality, says to take that axiom as the default implementation in instances and syntax systems. But we can punt on that for now. I'm not sure it's a good idea.

**Comment by epatters** [comment]
Resolved by #130.
